vertical-pod-autoscaler/enhancements/7862-cpu-startup-boost/README.md

```diff
@@ -0,0 +1,148 @@
+# AEP-7862: CPU Startup Boost
+
+<!-- toc -->
+- [AEP-7862: CPU Startup Boost](#aep-7862-cpu-startup-boost)
+  - [Summary](#summary)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+  - [Design Details](#design-details)
+    - [Workflow](#workflow)
+    - [API Changes](#api-changes)
+      - [Priority of `StartupBoost`](#priority-of-startupboost)
+      - [CPU startup boost and VPA autoscaling modes](#cpu-startup-boost-and-vpa-autoscaling-modes)
+    - [Validation](#validation)
+    - [Startup Boost and Readiness/Startup Probes](#startup-boost-and-readinessstartup-probes)
+    - [Failed in-place downsizes](#failed-in-place-downsizes)
+  - [Test Plan](#test-plan)
+  - [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Long application start time is a known problem for more traditional workloads
+running in containerized applications, especially Java workloads. This delay can
+negatively impact the user experience and overall application performance. One
+potential solution is to provide additional CPU resources to pods during their
+startup phase, but this can lead to waste if the extra CPU resources are not
+set back to their original values after the pods are ready.
+
+This proposal allows VPA to boost the CPU request and limit of containers during
+the pod startup and to scale the CPU resources back down when the pod is `Ready`,
+leveraging the [in-place pod resize Kubernetes feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources).
+
+
+### Goals
+
+* Allow VPA to boost the CPU request and limit of a pod's containers when the pod during the pod startup (from creation time until it becomes `Ready`.)
+* Allow VPA to scale pods down to the original CPU resource values, [in-place](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources), as soon as their [`Ready`](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) condition is true.
+
+### Non-Goals
+
+* Allow VPA to boost CPU resources of pods that are already [`Ready`](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions).
+* Allow VPA to boost CPU resources during startup of workloads that have not configured a [Readiness or a Startup probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).
+* Allow VPA to boost memory resources.
+
+## Proposal
+
+To extend [`ContainerResourcePolicy`](https://github.com/kubernetes/autoscaler/blob/0a34bf5d3a71b486bdaa440f1af7f8d50dc8e391/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L190) with a `StartupBoost.CPU.Factor` field, that will instruct VPA the factor by which to multiply the initial resource request and limit of the containers' targeted by the VPA object.
```

Comment:
Should the CPU factor be applied to the initial resource request and limit of the container, or on the VPA recommenders recommendation?  (this is similar to my previous comment)

Reply:
Is this assuming the startup is long enough for an actual recommendation to occur? I was assuming it was to the initial resource request and limit set on the container by the user.

Reply:
> Is this assuming the startup is long enough for an actual recommendation to occur? I was assuming it was to the initial resource request and limit set on the container by the user.

I had assumed two scenarios:
1. A brand new workload with no VPA history. I guess in this case whatever the VPA's "default" recommendation is would be the 'base' (I forget what happens on new workloads, I assume there is either no recommendation or a very young one), if that recommendation doesn't exist, then use the Pod spec resources
2. An old existing workload with a long history, in this case the base would be whatever the VPA's recommendation is

Reply:
I see, thanks for the explanation. I think it makes sense, though maybe there should be an option to use either method so that they don't run into the scenarios ray mentioned.

Like `startupBoost.boostMode=recommendation|initial` or something.

Reply:
A `StartupBoost.CPU.Value` option would be a great addition. 

For `StartupBoost.CPU.Factor`, what about we apply the boost factor to `max(request or limit in the spec, VPA recommendation)`, and then cap it to `maxAllowedCPU`? That would reduce the likelihood of  running into those issues and the API looks simpler than defining a `boostMode` like that?

Reply:
I re-wrote this part completely. Added `StartupBoost.CPU.Value` as an option and clarified that `StartupBoost.CPU.Factor` will be applied to either the existing VPA recommendation (if any) or the CPU resources in the pod spec, but the values will still respect `maxAllowedCPU`. 

Same for the scale down: we'll scale back down to the existing VPA recommendation (if any) or the CPU resources in the pod spec. 

This will also help us to avoid VPA doing a second actuation right after the pod is scaled down.